cvrp_metaheuristics/algorithm/tabu_search.py

```diff
@@ -0,0 +1,114 @@
+        return self.config.generations
+
+    @property
+    def mutation_type(self) -> Mutation:
+        return self.config.mutation_type
+
+    def solve(self) -> list[Result]:
+        results = []
+        for i in range(1, self.generations):
+            neighbours = self.search_neighbourhood()
+            best_neighbour = self.best_neighbour_not_in_tabu(neighbours)
+            # if all neighbours were already visited (are in tabu list) we're switching the neighbourhood drastically
+            # (random genotype)
+            if best_neighbour is None:
+                best_neighbour = self.change_neighbourhood()
+            self.current_best = best_neighbour
+            self.update_tabu(best_neighbour)
+            avg_fitness = round(sum([neighbour.fitness for neighbour in neighbours]) / len(neighbours), 2)
+            results.append(Result(self.current_best.fitness, avg_fitness, self.current_best.genotype))
+
+        return results
+
+    def init_individual(self) -> None:
+        if self.config.init_type == Initialization.RANDOM:
+            genotype = init_random_genotype(self.cvrp)
+        else:
+            genotype = init_greedy_genotype(self.cvrp, self.config)
+        self.current_best = Individual(genotype)
+        self.current_best.evaluate(self.cvrp)
+
+    def search_neighbourhood(self) -> list[Individual]:
+        neighbours: list[Individual] = []
+        for i in range(self.neighbourhood_size):
+            # TO CHECK: is it updated in place?
+            # neighbour_genotype = self.current_best.genotype[:]
+            # individual.mutate(neighbour_genotype, self.mutation_type)
+            neighbour_genotype = mutate(self.current_best.genotype[:], self.mutation_type)
+            neighbour = Individual(neighbour_genotype)
+            neighbour.evaluate(self.cvrp)
+            neighbours.append(neighbour)
+        return neighbours
+
+    def best_neighbour_not_in_tabu(self, neighbours: list[Individual]) -> Individual | None:
+        best_neighbour = None
+        best_neighbour_fitness = float('inf')
+        for i in range(1, len(neighbours)):
+            if (neighbours[i].fitness < best_neighbour_fitness
+                    and hash(tuple(neighbours[i].genotype)) not in self.tabu_set):
+                best_neighbour_fitness = neighbours[i].fitness
+                best_neighbour = neighbours[i]
+        return best_neighbour
+
+    def update_tabu(self, best_neighbour) -> None:
+        genotype_hash = hash(tuple(best_neighbour.genotype))
+        self.tabu_set.add(genotype_hash)
+        self.tabu_list.append(genotype_hash)
+        if len(self.tabu_list) > self.config.tabu_list_size:
+            self.tabu_set.remove(self.tabu_list.pop(0))
+
+    def change_neighbourhood(self) -> Individual:
+        genotype = init_random_genotype(self.cvrp)
+        while hash(tuple(genotype)) in self.tabu_set:
+            random.shuffle(genotype)
+        i = Individual(genotype)
+        i.evaluate(self.cvrp)
```

Comment:
This a tangent, but you could add an `evaluate_genotype_fitness` method to `Cvrp` that `Individual.evaluate` calls and decorate that with `@cache`, that way you can avoid writing your own custom caches. You would probably have to turn the genotypes into tuples of integers, but kinda feels appropriate anyway? Is it important that they're mutable?

Reply:
Another tangent, what's up with the apostrophes in the argument type hint here:

```def evaluate(self, cvrp: 'Cvrp')```

Reply:
Some of the operators make updates in place, so for now I'd like to keep the list as underlying data structure.
Tuple would be more in alignment with paradigms of functional programming, right? (no side effects)

Reply:
--  Tuple would be more in alignment with paradigms of functional programming, right? (no side effects)

Kinda, I mean, they stop you accidentally modifying the list somewhere else without realizing that it'll have an effect here (you know, replacing any mutable object with an immutable counterpart). In this case, I got the feeling that there wasn't any point in having genotype be mutable (do you ever wanna modify it in place?). If not, then inmutable is always safer. But in this particular case, all I had in mind is that tuples are hashable (while anything mutable is not), which you'll need to use the `@cache` decorator.